### PR TITLE
handle nonRoot -> root FG transition

### PR DESCRIPTION
**Pull request by tiraboschi**
**What this PR does / why we need it**:
Up to version v0.58, Kubevirt was consuming
a featuregate named nonRoot, with
v0.59 it got renamed as Root with the opposite
logic and so we implemented
https://github.com/kubevirt/hyperconverged-cluster-operator/pull/2273
as an internal workaround for the time being.

Let's now align with Kubevirt featureGate name
for the sake of consistency with a proper deprecation
plan.

`nonRoot` is still exposed on HCO but it's
deprecated and a new `Root` FG is introduced.

The mutating webhook will take care
of the conversion.

A Common Expression Language rule on the CRD
will enforce the consistency of the two antithetic
values.

In the next release `nonRoot` will be removed.

Adding unit and e2e tests.

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=2174859

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly

**Jira Ticket**:

```jira-ticket
https://issues.redhat.com/browse/CNV-26376
```

**Release note**:

```release-note
deprecate nonRoot FG and introduce root FG
```

### CODE REVIEW

File: pkg/webhooks/mutator/hyperConvergedMutator.go [hunk 1]
```diff
@@ -84,6 +84,18 @@ func (hcm *HyperConvergedMutator) mutateHyperConverged(_ context.Context, req ad
 		}
 	}
 
+	if hc.Spec.FeatureGates.Root == nil {
+		value := false
+		if hc.Spec.FeatureGates.NonRoot != nil {
+			value = !*hc.Spec.FeatureGates.NonRoot
```
Comment (nunnatsa) [inline pkg/webhooks/mutator/hyperConvergedMutator.go:90]: Maybe we want to remove the old field as well?

File: tests/func-tests/nonroot_root_test.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+				hc = tests.UpdateHCORetry(ctx, cli, hc)
+				Expect(hc.Spec.FeatureGates.NonRoot).To(Equal(expectedNonRootFG))
+				Expect(hc.Spec.FeatureGates.Root).To(Equal(expectedRootFG))
+			} else {
+				Eventually(func() error {
+					hc := tests.GetHCO(ctx, cli)
+					hc.Spec.FeatureGates.NonRoot = nonRootFG
+					hc.Spec.FeatureGates.Root = rootFG
+					hc, err = tests.UpdateHCO(ctx, cli, hc)
+					return err
```
Comment (nunnatsa) [inline tests/func-tests/nonroot_root_test.go:66]: Do you want to check the KubeVirt CR as well?

File: api/v1beta1/hyperconverged_types.go [hunk 3]
```diff
@@ -332,15 +333,22 @@ type HyperConvergedFeatureGates struct {
 	// +default=false
 	DeployKubeSecondaryDNS *bool `json:"deployKubeSecondaryDNS,omitempty"`
 
-	// TODO: deprecate NonRoot and introduce Root FG with opposite semantic
-	// and a proper conversion logic
+	// TODO: remove NonRoot in the next release
 
-	// Enables rootless virt-launcher.
+	// Enables rootless virt-launcher. DEPRECATED: please use the root FG.
```
[the remark this answers is not in the document]
Comment (tiraboschi) [reply api/v1beta1/hyperconverged_types.go:338]: done, thanks

File: api/v1beta1/hyperconverged_types.go [hunk 4]
```diff
@@ -332,15 +333,22 @@ type HyperConvergedFeatureGates struct {
...
 	// +optional
-	// +kubebuilder:default=true
-	// +default=true
 	NonRoot *bool `json:"nonRoot,omitempty"`
 
+	// TODO: skipping any default on root here to avoid any
+	// conflict during CRD updates.
+	// The default will be set by a mutating webhook.
+	// Enable back a default here when we can finally get rid of
+	// NonRoot
```
[the remark this answers is not in the document]
Comment (tiraboschi) [reply api/v1beta1/hyperconverged_types.go:346]: the TODO is about enabling back the default here when we will be able to get rid of the deprecated FG and so the conversion logic in the mutating webhook

File: tests/func-tests/nonroot_root_test.go (hunk 2 above)
Comment (tiraboschi) [reply tests/func-tests/nonroot_root_test.go:80]: done

**Pull request merged**